tempesta_fw/client.c

```diff
@@ -28,122 +28,161 @@
 #include "connection.h"
 #include "log.h"
 #include "procfs.h"
+#include "tdb.h"
+#include "lib/str.h"
+#include "lib/common.h"
 
-#define CLI_HASH_BITS	17
-#define CLI_HASH_SZ	(1 << CLI_HASH_BITS)
+static struct {
+	unsigned int	db_size;
+	const char	*db_path;
+	unsigned int	lifetime;
+} client_cfg __read_mostly;
 
 typedef struct {
-	struct hlist_head	list;
-	spinlock_t		lock;
-} CliHashBucket;
+	TfwClient	cli;
+	time_t		expires;
+	spinlock_t	lock;
+	atomic_t	users;
+} TfwClientEntry;
```

Comment:
Explanatory comments for structure fields are appreciated, especially - for `lock` field (which protect against `tfw_client_obtain/tfw_client_put` race condition), since we have now rather complicated synchronization case in `tfw_client_obtain()` -> `tdb_rec_get_alloc()` with three separate nested locks with different protected resources.
In this context - it would also be nice to add comment for global `get_alloc_lock` (in tempesta_db/core/main.c), which provides now atomic get/alloc TDB operation and protection against races during instances' constructor execution.